### PR TITLE
fix: textarea value not saved/rendered

### LINKED ISSUE #388: textarea value not rendered
image

Error rendering textarea values:

Data test example:

{
  "id": "f6c6ebaf-5923-48d4-bea6-27afbcb9c0ce",
  "stages": {
    "f3f9328c-f022-453b-bddb-f1b5905d9010": {
      "children": [
        "319b03c3-1d8d-4f45-973c-244800b3f5ab",
        "985a993d-aca8-4bf7-a687-bddf70ed0c5b",
        "43a992f6-342b-4a89-a3b5-43b0bbd72229"
      ],
      "id": "f3f9328c-f022-453b-bddb-f1b5905d9010"
    }
  },
  "rows": {
    "319b03c3-1d8d-4f45-973c-244800b3f5ab": {
      "config": {
        "fieldset": false,
        "legend": "",
        "inputGroup": false
      },
      "children": [
        "a37b80e0-8f24-49d5-bd67-0f897558e9d0"
      ],
      "className": [
        "formeo-row"
      ],
      "id": "319b03c3-1d8d-4f45-973c-244800b3f5ab"
    },
    "985a993d-aca8-4bf7-a687-bddf70ed0c5b": {
      "config": {
        "fieldset": false,
        "legend": "",
        "inputGroup": false
      },
      "children": [
        "121ba22e-7da9-4b25-b77c-0c8882866006"
      ],
      "className": [
        "formeo-row"
      ],
      "id": "985a993d-aca8-4bf7-a687-bddf70ed0c5b"
    },
    "43a992f6-342b-4a89-a3b5-43b0bbd72229": {
      "config": {
        "fieldset": false,
        "legend": "",
        "inputGroup": false
      },
      "children": [
        "7ba8089a-91eb-4bf4-b020-b1d1f1964c51"
      ],
      "className": [
        "formeo-row"
      ],
      "id": "43a992f6-342b-4a89-a3b5-43b0bbd72229"
    }
  },
  "columns": {
    "a37b80e0-8f24-49d5-bd67-0f897558e9d0": {
      "config": {
        "width": "100%"
      },
      "children": [
        "2f521f05-2b42-40c8-a333-bfe8bf1c3e79"
      ],
      "className": [
        "formeo-column"
      ],
      "id": "a37b80e0-8f24-49d5-bd67-0f897558e9d0"
    },
    "121ba22e-7da9-4b25-b77c-0c8882866006": {
      "config": {
        "width": "100%"
      },
      "children": [
        "8423432e-ec7f-4bfb-aeb0-0767b81b8262"
      ],
      "className": [
        "formeo-column"
      ],
      "id": "121ba22e-7da9-4b25-b77c-0c8882866006"
    },
    "7ba8089a-91eb-4bf4-b020-b1d1f1964c51": {
      "config": {
        "width": "100%"
      },
      "children": [
        "0bbb9ebe-ac86-4326-8400-7fd1f86d8db9"
      ],
      "className": [
        "formeo-column"
      ],
      "id": "7ba8089a-91eb-4bf4-b020-b1d1f1964c51"
    }
  },
  "fields": {
    "2f521f05-2b42-40c8-a333-bfe8bf1c3e79": {
      "tag": "h1",
      "attrs": {
        "tag": [
          {
            "label": "H1",
            "value": "h1",
            "selected": true
          },
          {
            "label": "H2",
            "value": "h2",
            "selected": false
          },
          {
            "label": "H3",
            "value": "h3",
            "selected": false
          },
          {
            "label": "H4",
            "value": "h4",
            "selected": false
          }
        ],
        "className": ""
      },
      "config": {
        "label": "Encabezamiento",
        "hideLabel": true,
        "editableContent": true,
        "controlId": "html.header"
      },
      "content": "<span style=\"font-family: Inter, &quot;Helvetica Neue&quot;, Arial, -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, &quot;Noto Sans&quot;, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;;\">Entrevista&nbsp;</span>",
      "action": [],
      "id": "2f521f05-2b42-40c8-a333-bfe8bf1c3e79"
    },
    "8423432e-ec7f-4bfb-aeb0-0767b81b8262": {
      "tag": "textarea",
      "config": {
        "label": "<span style=\"font-family: Inter, &quot;Helvetica Neue&quot;, Arial, -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, &quot;Noto Sans&quot;, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;;\">1. ¿Por qué te interesa formar parte de nuestro equipo en Dental OralCare?</span>",
        "controlId": "textarea"
      },
      "attrs": {
        "required": false,
        "value": "sss"
      },
      "id": "8423432e-ec7f-4bfb-aeb0-0767b81b8262",
      "content": ""
    },
    "0bbb9ebe-ac86-4326-8400-7fd1f86d8db9": {
      "tag": "input",
      "attrs": {
        "required": false,
        "type": "text",
        "className": "",
        "value": "jaider"
      },
      "config": {
        "label": "Text Input",
        "controlId": "text-input"
      },
      "id": "0bbb9ebe-ac86-4326-8400-7fd1f86d8db9"
    }
  }
}

_Originally posted by @JordanGarciaDev in https://github.com/Draggable/formeo/issues/386#issuecomment-2543174262_

### PR DESCRIPTION
**Pull request by kevinchappell**
resolves #388

### CODE REVIEW

File: src/lib/js/components/controls/form/textarea.js [hunk 1]
```diff
@@ -8,26 +8,12 @@ class TextAreaControl extends Control {
...
-      //       }
-      //       target.removeEventListener('mouseup', onMouseup);
-      //     };
-      //     target.addEventListener('mouseup', onMouseup);
-      //   }
-      // },
+      // actions here will be applied to the preview in the editor
+      action: {
+        input: function ({ target: { value } }) {
+          this.setData?.('value', value)
```
Comment (kevinchappell) [inline src/lib/js/components/controls/form/textarea.js:14]: `this` will be the Field instance when called from editor preview

File: src/lib/js/components/controls/index.js [hunk 2]
```diff
@@ -379,7 +375,14 @@ export class Controls {
     this.container = container
     this.groupOrder = unique(groupOrder.concat(['common', 'html', 'layout']))
     this.options = options
-    return Promise.all(this.registerControls([...defaultElements, ...elements]))
+
+    const layoutControls = await import('./layout/index.js')
```
Comment (kevinchappell) [inline src/lib/js/components/controls/index.js:379]: using dynamic import to prevent circular dependency found when writing tests for controls

File: src/lib/js/components/fields/field.js [hunk 3]
```diff
@@ -130,13 +130,15 @@ export default class Field extends Component {
     }
   }
 
+  setData = (path, value) => {
```
Comment (kevinchappell) [inline src/lib/js/components/fields/field.js:133]: shortcut for setting data without updating the preview or edit panels

File: src/lib/js/components/fields/field.js [hunk 4]
```diff
@@ -310,7 +312,10 @@ export default class Field extends Component {
     const prevData = clone(this.data)
     const { action = {} } = controls.get(prevData.config.controlId)
     prevData.id = `prev-${this.id}`
-    prevData.action = action
+    prevData.action = Object.entries(action).reduce((acc, [key, value]) => {
+      acc[key] = value.bind(this)
```
Comment (kevinchappell) [inline src/lib/js/components/fields/field.js:316]: not sure about this one but it binds the Field instance so we have access to `setData` from the component class.

**Comment by JordanGarciaDev** [comment]
image

You changed the logic and the meaning of the textarea as the rest of the inputs are created

The value was already being stored like the other inputs, the problem was that when the form was rendered it didn't grab the value of the attrs

Now all the logic changes, and you have to add a lot more code to just condition the textarea, the idea was that just as it renders the input text, email, etc

it would also grab the value from the attrs array, and not leave it aside.

In other words, it was going to take the value of attrs and not create another value outside of attrs

**Comment by kevinchappell** [comment]
i see what you mean. will take another look

**Comment by kevinchappell** [comment]
`textarea` doesn't take a value attribute the same way as a standard `input` tag. i can add additional logic to `dom.create` to work around this but I think the better option is logic on the control class as it was implemented.
[textarea MDN image](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea)


essentially, the HTMLTextAreaElement has a setter that allows `textarea.value = 'whatever'` to set the innerText of the element which is why the current implementation works. Is the current implementation not working? I only made changes to the textarea control class, you should not need to provide any additional configuration for the textarea to function as expected. Is there an issue storing the `value` at the root of the element data?

**Comment by kevinchappell** [comment]
[i removed the custom input handler](https://github.com/Draggable/formeo/pull/392/files) on the textarea class in favor of the built-in preview handler on the Field class. this still uses a root level assignment of content instead of value as `value` is not a valid attribute of `textarea` and making it work like `input` would be a workaround in itself.
